text/0000-cargo-features2.md

```diff
@@ -0,0 +1,519 @@
+Feedback on desired use cases for feature information will help define the
+solution. A possible alternative is to stabilize the [`--unit-graph`] flag,
+which exposes Cargo's internal graph structure, which accurately indicates the
+actual dependency relationships and uses the new feature resolver.
+
+For non-parseable output, `cargo tree` will show features from the new
+resolver.
+
+[`--unit-graph`]: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#unit-graph
+
+# Drawbacks
+
+There are a number of drawbacks to this approach:
+
+* In some situations, dependencies will be built multiple times where they
+  were previously only built once. This causes two problems: increased build
+  times, and potentially broken builds when transitioning to the new resolver.
+  It is intended that if the user wants to build a dependency once that now
+  has non-unified features, they will need to add feature declarations within
+  their dependencies so that they once again have the same features. The
+  `cargo tree` command has been added to help the user identify and remedy
+  these situations. `cargo tree -d` will expose dependencies that are built
+  multiple times, and the `-e features` flag can be used to see which packages
+  are enabling which features.
+
+  Unfortunately the error message is not very clear when a feature that was
+  previously assumed to be enabled is no longer enabled. Typically these
+  appear in the form of unresolved paths. In testing so far, this has come up
+  occasionally, but is usually fairly easy to identify what is wrong. Once
+  more of the ecosystem starts using the new resolver, these errors should
+  become less frequent.
+
+* Feature unification with dev-dependencies being a global decision can result
+  in some artifacts including features that may not be desired. For example, a
+  project with a binary and a shared dependency that is used as a
+  dev-dependency and a normal dependency. When running `cargo test` the binary
+  will include the shared dev-dependency features. Compare this to a normal
+  `cargo build --bin name`, where the binary will be built without those
+  features. This means that if you are testing a binary with an integration
+  test, you end up not testing the same thing as what is normally built.
+  Changing this has significant drawbacks. Cargo's dependency graph
+  construction will require fundamental changes to support this scenario.
+  Additionally, it has a high risk that will cause increased build times for
+  many projects that aren't affected or don't care that it may have slightly
+  different features enabled.
+
+* This adds complexity to Cargo, and adds boilerplate to `Cargo.toml`. It can
+  also be confusing when switching between projects that use different
+  settings. It is intended in the future that new resolver will become the
+  default via the "edition" declaration. This will remove the extra
```

Comment:
`edition` is a non-global choice and there workspaces generally do not have the edition annotation. Does this mean that workspaces are doomed to forever using `resolver` annotation and/or using the old resolver?

Reply:
We are planning to add the `edition` field to the `[workspace]` table. That is still a bit up in the air, since we haven't really defined what that will do (would it be the default for all workspace members?).  Issue https://github.com/rust-lang/cargo/issues/5784 is tracking that.  One of my concerns is that with [RFC 2906](https://github.com/rust-lang/rfcs/blob/master/text/2906-cargo-workspace-deduplicate.md), that inheriting package metadata will require specifying `{ workspace = true }` and that ends up creating more boilerplate in each package.  But making it implicitly inherit the edition would be inconsistent with the other fields.